stern/watch.go

```diff
@@ -79,79 +64,19 @@ func Watch(ctx context.Context, i v1.PodInterface, podFilter *regexp.Regexp, exc
 					continue
 				}
 
-				if !podFilter.MatchString(pod.Name) {
-					continue
-				}
-
-				if excludePodFilter != nil && excludePodFilter.MatchString(pod.Name) {
-					continue
-				}
-
 				switch e.Type {
 				case watch.Added, watch.Modified:
-					var statuses []corev1.ContainerStatus
-					statuses = append(statuses, pod.Status.ContainerStatuses...)
-					if initContainers {
-						statuses = append(statuses, pod.Status.InitContainerStatuses...)
-					}
-
-					if ephemeralContainers {
-						statuses = append(statuses, pod.Status.EphemeralContainerStatuses...)
-					}
-
-					for _, c := range statuses {
-						if !containerFilter.MatchString(c.Name) {
-							continue
-						}
-						if containerExcludeFilter != nil && containerExcludeFilter.MatchString(c.Name) {
-							continue
-						}
-						t := &Target{
-							Node:      pod.Spec.NodeName,
-							Namespace: pod.Namespace,
-							Pod:       pod.Name,
-							Container: c.Name,
-						}
-
-						containerStateMatched := false
-						for _, containerState := range containerStates {
-							if containerState.Match(c.State) {
-								containerStateMatched = true
-								break
-							}
-						}
+					filter.visit(pod, func(t *Target, containerStateMatched bool) {
 						if containerStateMatched {
 							added <- t
 						} else {
 							removed <- t
 						}
-					}
+					})
 				case watch.Deleted:
-					var containerNames []string
-					containerNames = append(containerNames, getContainerNames(pod.Spec.Containers)...)
-					if initContainers {
-						containerNames = append(containerNames, getContainerNames(pod.Spec.InitContainers)...)
-					}
-
-					if ephemeralContainers {
-						containerNames = append(containerNames, getEphemeralContainerNames(pod.Spec.EphemeralContainers)...)
-					}
-
-					for _, cn := range containerNames {
-						if !containerFilter.MatchString(cn) {
-							continue
-						}
-						if containerExcludeFilter != nil && containerExcludeFilter.MatchString(cn) {
-							continue
-						}
-
-						removed <- &Target{
-							Node:      pod.Spec.NodeName,
-							Namespace: pod.Namespace,
-							Pod:       pod.Name,
-							Container: cn,
-						}
-					}
+					filter.visit(pod, func(t *Target, containerStateMatched bool) {
+						removed <- t
+					})
```

Comment:
Containers without their statuses are never added, so now we get container names from container statuses in the same manner as handling Added/Modified events to simplify the logic.